enclave/core/sgx/once.c

```diff
@@ -5,34 +5,39 @@
 #include <openenclave/internal/thread.h>
 #include <openenclave/internal/utils.h>
 
+#define SPINLOCK_NOT_USED 0
+#define SPINLOCK_ACTIVELY_HELD 1
+#define SPINLOCK_USED 2
+
 oe_result_t oe_once(oe_once_t* once, void (*func)(void))
 {
     if (!once)
         return OE_INVALID_PARAMETER;
 
     /* Double checked locking (DCLP). */
-    oe_once_t o = *once;
-
     /* DCLP Acquire barrier. */
     OE_ATOMIC_MEMORY_BARRIER_ACQUIRE();
-    if (o == 0)
+    if (*once != SPINLOCK_USED)
     {
-        static oe_spinlock_t _lock = OE_SPINLOCK_INITIALIZER;
-
-        oe_spin_lock(&_lock);
-
-        if (*once == 0)
+        oe_once_t retval = __sync_val_compare_and_swap(
+            once, SPINLOCK_NOT_USED, SPINLOCK_ACTIVELY_HELD);
+        if (retval == SPINLOCK_NOT_USED)
         {
             if (func)
                 func();
 
-            /* DCLP Release barrier. */
-            OE_ATOMIC_MEMORY_BARRIER_RELEASE();
-            *once = 1;
+            *once = SPINLOCK_USED;
+        }
+        else if (retval == SPINLOCK_ACTIVELY_HELD)
+        {
+            while (__sync_val_compare_and_swap(
```

Comment:
"Another thread is invoking the function. Wait for that thread to finish the invocation and mark the once variable to FUNC_INVOKED".